### DISCUSSION TITLE
Why does my dbt target directory keep making new directories for each dbt run?

### DISCUSSION THREAD

**Discussion by dagsir** [dagster-io/dagster#18235] [Q&A] [answered]
Hi!
We use dagster-dbt a lot, and we use it in a separate code location.
We load many dbt models individually (to assign them partitions) in a function called `materialize_&lt;dbt_model_name&gt;`, and every time you start a run for a single dbt-asset, Dagster generates a subfolder in the dbt "target" directory:
`materialize_&lt;dbt_model_name&gt;-&lt;run-id&gt;-&lt;some_other_random_id&gt;`

What seems weird to me, is that this folder is generated and filled with new dbt artifacts (manifes.json, ...). If you run dbt manually (dbt build --select &lt;some_model&gt;), it uses the existing manifest.json in the target folder.
(1) is there a reason for this behaviour?

But the problem is rather, that when developing a lot and executing several runs in Dagster, you end up with many of these subfolders in the target folder. To keep the target folder clean, I am currently deleting those folders via a function in the _init.py_ of the code location. However, Dagster apparently executes this _init.py_ not only when reloading the code location, but rather more often, about every 1-2 minutes. This leads to the problem, that the subfolders in target sometimes get deleted even before dagster finished a run. This then sometimes leads to errors because dagster cannot find the folder anymore. I rather just want to clean the target folder only when actually wanted, such as with a reload of the code location.
(2) How can I achieve this? I don't want the dbt-developers to always think about manually cleaning the target folder...

---

*The question was originally asked in [Dagster Slack](https://dagster.slack.com/archives/C04CW71AGBW/p1700583730389829?thread_ts=1700583730.389829&cid=C04CW71AGBW).*

**Comment by tacastillo** [comment, marked as answer]
From @OwenKephart:

dagster-dbt creates a separate target folder for each invocation to avoid issues with multiple concurrent dbt invocations (in separate processes) writing to the same output files. Generally, the manifest.json will be copied over into that temporary directory, so it will not actually be regenerated on each dbt invocation.
There's unfortunately not really a great way of cleaning these up without adding an extra line at the end of your @dbt_assets function (the DbtCliInvocation object has a target_path attribute that you can use to grab the folder to delete), but depending on your setup this may not involve dbt developers needing to do anything manual themselves (this could just be part of your framework).

**Comment by cmpadden** [comment]
For future reference, @rexledesma has pointed out that it's possible to overwrite the `target/` directory to prevent it from growing.

https://discuss.dagster.io/t/16075662/hi-i-made-dagster-parse-dbt-project-every-time-it-s-launched#4bcafc64-de69-4a26-93e5-38a67f98fade

> you could add the following to your dbt parse, so that your manifest is always created in the target/ directory, overwriting previous manifests.

```python
dbt_manifest_path = (
    dbt_resource.cli(
        ["--quiet", "parse"],
        target_path=Path("target)
    )
    .wait()
    .target_path.joinpath("manifest.json")
)
```
